source/encoding.c

```diff
@@ -0,0 +1,316 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/encoding.h>
+
+static const char *hex_chars = "0123456789abcdef";
+
+static const uint8_t base64_sentinal_value = 0xff;
+static const char base64_encoding_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+static uint8_t base64_decoding_table[256];
+static int8_t base64_decoding_table_init = 0;
+
+static const uint8_t bytes_7 = 7 * 8;
+static const uint8_t bytes_6 = 6 * 8;
+static const uint8_t bytes_5 = 5 * 8;
+static const uint8_t bytes_4 = 4 * 8;
+static const uint8_t bytes_3 = 3 * 8;
+static const uint8_t bytes_2 = 2 * 8;
+static const uint8_t bytes_1 = 1 * 8;
+static const uint8_t mask = 0xff;
+
+
+int aws_hex_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size) {
```

Comment:
Declaring these pointers as `type *restrict varname` will help the compiler in these kinds of buffer-heavy operations, by declaring that the input and output cannot alias each other (nor, for that matter, can they alias the encoding table). If the compiler is sufficiently clever it may also enable vectorization.